crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1401,6 +1406,112 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Returns true if this type and `other` are "exactly the same".
+    ///
+    /// Two types are "exactly the same" when:
+    ///
+    /// * Both are either `Any` or `Unknown`, or
+    /// * If they are not tuples, generics, unions or intersections:
+    ///     * They share the same internal representation.
+    /// * If they are tuples:
+    ///     * They have the same length, and
+    ///     * Each element pair are "exactly the same".
+    /// * If they are generics:
+    ///     * The "origins" are "exactly the same".
+    ///     * The argument types are "exactly the same" when compared as tuples.
+    /// * If they are unions:
+    ///     * They have the same length, and
+    ///     * For each element of the first type,
+    ///       there exist one element of the second
+    ///       such that they are "exactly the same";
+    ///       each element must not be paired more than once.
+    /// * If they are intersections:
+    ///     * The positive parts are "exactly the same" when compared as unions.
+    ///     * The negative parts are "exactly the same" when compared as unions.
+    ///
+    /// Note: `Todo != Todo`.
+    ///
+    /// This powers the `assert_type()` directive.
+    pub(crate) fn is_equals_to(self, db: &'db dyn Db, other: Type<'db>) -> bool {
+        let equal = |(first, second): (&Type<'db>, &Type<'db>)| first.is_equals_to(db, *second);
+
+        match (self, other) {
+            (Type::Todo(_), Type::Todo(_)) => false,
+
+            (Type::Any | Type::Unknown, Type::Any | Type::Unknown)
+            | (Type::Never, Type::Never)
+            | (Type::AlwaysTruthy, Type::AlwaysTruthy)
+            | (Type::AlwaysFalsy, Type::AlwaysFalsy)
+            | (Type::LiteralString, Type::LiteralString) => true,
+
+            (Type::KnownInstance(first), Type::KnownInstance(second)) => first == second,
+
+            (Type::FunctionLiteral(first), Type::FunctionLiteral(second)) => first == second,
+
+            (Type::ModuleLiteral(first), Type::ModuleLiteral(second)) => {
+                first.module(db) == second.module(db)
+            }
+
+            (Type::IntLiteral(first), Type::IntLiteral(second)) => first == second,
+            (Type::BooleanLiteral(first), Type::BooleanLiteral(second)) => first == second,
+            (Type::StringLiteral(first), Type::StringLiteral(second)) => {
+                first.value(db).as_ref() == second.value(db).as_ref()
+            }
+            (Type::BytesLiteral(first), Type::BytesLiteral(second)) => {
+                first.value(db).as_ref() == second.value(db).as_ref()
+            }
+            (Type::SliceLiteral(first), Type::SliceLiteral(second)) => {
+                first.as_tuple(db) == second.as_tuple(db)
+            }
+
+            (Type::ClassLiteral(first), Type::ClassLiteral(second)) => first.class == second.class,
+            (Type::Instance(first), Type::Instance(second)) => first.class == second.class,
```

Comment:
For all fully-static types, I think we could just defer to `is_equivalent_to`? And/or we could have a fast-path at the top of this method for `if self == other { return true; }`, similar to `is_equivalent_to`. Either one would eliminate the need for many of these cases.